dvc/remote/local/link_decorator.py

```diff
@@ -0,0 +1,56 @@
+import logging
+import time
+
+from dvc.config import Config
+from dvc.utils import boxify
+
+
+logger = logging.getLogger(__name__)
+
+
+class SlowLinkDetectorDecorator(object):
+    LINKING_TIMEOUT_SECONDS = 10.0
+
+    def __init__(self, method):
+        self.method = method
+
+    def __call__(self, remote_local, *args, **kwargs):
+        start = time.time()
+        result = self.method(remote_local, *args, **kwargs)
+        execution_time_seconds = time.time() - start
+
+        if execution_time_seconds >= self.LINKING_TIMEOUT_SECONDS:
+
+            cache_config = remote_local.repo.config.config.get(
```

Comment:
Your decorator class doesn't have to parse configs. You could simply decide in slow_link_aware::call whether or not you should wrap the call in this decorator or if you just pass it through. That way it will be much easier to read and much easier to unit test, since you won't have to mock a lot of stuff.